homeassistant/components/sensor/statistics.py

```diff
@@ -0,0 +1,153 @@
+"""
+Support for statistics for sensor values.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.statistics/
+"""
+import logging
+import statistics
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_NAME, CONF_ENTITY_ID, STATE_UNKNOWN, ATTR_UNIT_OF_MEASUREMENT)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_state_change
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_MIN_VALUE = 'min_value'
+ATTR_MAX_VALUE = 'max_value'
+ATTR_COUNT = 'count'
+ATTR_MEAN = 'mean'
+ATTR_MEDIAN = 'median'
+ATTR_VARIANCE = 'variance'
+ATTR_STANDARD_DEVIATION = 'standard_deviation'
+ATTR_SAMPLING_SIZE = 'sampling_size'
+ATTR_TOTAL = 'total'
+
+CONF_SAMPLING_SIZE = 'sampling_size'
+DEFAULT_NAME = 'Stats'
+DEFAULT_SIZE = 0
+ICON = 'mdi:calculator'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_SAMPLING_SIZE, default=DEFAULT_SIZE): cv.positive_int,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the Statistics sensor."""
+    entity_id = config.get(CONF_ENTITY_ID)
+    name = config.get(CONF_NAME)
+    sampling_size = config.get(CONF_SAMPLING_SIZE)
+
+    add_devices([StatisticsSensor(hass, entity_id, name, sampling_size)])
+
+
+# pylint: disable=too-many-instance-attributes
+class StatisticsSensor(Entity):
+    """Representation of a Statistics sensor."""
+
+    def __init__(self, hass, entity_id, name, sampling_size):
+        """Initialize the Statistics sensor."""
+        self._hass = hass
+        self._entity_id = entity_id
+        self.is_binary = True if self._entity_id.split('.')[0] == \
+            'binary_sensor' else False
+        if not self.is_binary:
+            self._name = '{} {}'.format(name, ATTR_MEAN)
+        else:
+            self._name = '{} {}'.format(name, ATTR_COUNT)
+        self._sampling_size = sampling_size
+        self._unit_of_measurement = None
+        self.states = []
+        self.median = self.mean = self.variance = self.standard_deviation = 0
+        self.min = self.max = self.total = self.count = 0
+        self.update()
+
+        def calculate_sensor_state_listener(entity, old_state, new_state):
+            """Called when the sensor changes state."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+
+            try:
+                state = float(new_state.state)
+
+                if self._sampling_size is 0:
+                    self.states.append(state)
+                elif len(self.states) is self._sampling_size:
+                    del self.states[0]
```

Comment:
Use a [`Deque`](https://docs.python.org/3/library/collections.html#collections.deque) for optimal performance here and built-in lenght support.